File: frontend/lib/models/friend_request.dart
import 'user.dart';

class FriendRequest {
  final String id;
  final User sender;
  final User receiver;
  final String status;
  final String message;
  final DateTime? respondedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  FriendRequest({
    required this.id,
    required this.sender,
    required this.receiver,
    required this.status,
    this.message = '',
    this.respondedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory FriendRequest.fromJson(Map<String, dynamic> json) {
    return FriendRequest(
      id: json['_id'] ?? json['id'] ?? '',
      sender: User.fromJson(json['sender']),
      receiver: User.fromJson(json['receiver']),
      status: json['status'] ?? 'pending',
      message: json['message'] ?? '',
      respondedAt: json['respondedAt'] != null 
          ? DateTime.parse(json['respondedAt']) 
          : null,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'sender': sender.toJson(),
      'receiver': receiver.toJson(),
      'status': status,
      'message': message,
      'respondedAt': respondedAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  FriendRequest copyWith({
    String? id,
    User? sender,
    User? receiver,
    String? status,
    String? message,
    DateTime? respondedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return FriendRequest(
      id: id ?? this.id,
      sender: sender ?? this.sender,
      receiver: receiver ?? this.receiver,
      status: status ?? this.status,
      message: message ?? this.message,
      respondedAt: respondedAt ?? this.respondedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  bool get isPending => status == 'pending';
  bool get isAccepted => status == 'accepted';
  bool get isDeclined => status == 'declined';
  bool get isCancelled => status == 'cancelled';

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is FriendRequest && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'FriendRequest(id: $id, sender: ${sender.username}, receiver: ${receiver.username}, status: $status)';
  }
}

File: backend/src/models/FriendRequest.js
const mongoose = require('mongoose');

const friendRequestSchema = new mongoose.Schema({
  sender: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  receiver: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  status: {
    type: String,
    enum: ['pending', 'accepted', 'declined', 'cancelled'],
    default: 'pending'
  },
  message: {
    type: String,
    trim: true,
    maxlength: [200, 'Friend request message cannot exceed 200 characters'],
    default: ''
  },
  respondedAt: {
    type: Date,
    default: null
  }
}, {
  timestamps: true
});

// Index for faster queries
friendRequestSchema.index({ sender: 1, receiver: 1 });
friendRequestSchema.index({ receiver: 1, status: 1 });
friendRequestSchema.index({ sender: 1, status: 1 });
friendRequestSchema.index({ createdAt: -1 });

// Ensure unique friend request between two users
friendRequestSchema.index(
  { sender: 1, receiver: 1 },
  { 
    unique: true,
    partialFilterExpression: { status: 'pending' }
  }
);

// Virtual for request age
friendRequestSchema.virtual('requestAge').get(function() {
  return Date.now() - this.createdAt.getTime();
});

// Method to accept friend request
friendRequestSchema.methods.accept = async function() {
  const User = require('./User');
  
  // Update request status
  this.status = 'accepted';
  this.respondedAt = new Date();
  await this.save();

  // Add each user to the other's friends list
  await User.findByIdAndUpdate(
    this.sender,
    { $addToSet: { friends: this.receiver } }
  );
  
  await User.findByIdAndUpdate(
    this.receiver,
    { $addToSet: { friends: this.sender } }
  );

  return this;
};

// Method to decline friend request
friendRequestSchema.methods.decline = async function() {
  this.status = 'declined';
  this.respondedAt = new Date();
  return await this.save();
};

// Method to cancel friend request
friendRequestSchema.methods.cancel = async function() {
  this.status = 'cancelled';
  this.respondedAt = new Date();
  return await this.save();
};

// Static method to check if friend request exists
friendRequestSchema.statics.existsBetween = async function(senderId, receiverId) {
  return await this.findOne({
    $or: [
      { sender: senderId, receiver: receiverId, status: 'pending' },
      { sender: receiverId, receiver: senderId, status: 'pending' }
    ]
  });
};

// Static method to check if users are already friends
friendRequestSchema.statics.areFriends = async function(userId1, userId2) {
  const User = require('./User');
  const user = await User.findById(userId1).select('friends');
  return user && user.friends.includes(userId2);
};

module.exports = mongoose.model('FriendRequest', friendRequestSchema);
